From 40cc0f27774ee2022b4e1ea8adaf34d882079017 Mon Sep 17 00:00:00 2001
From: Pradeep M <pradeep.m@intrinsyc.com>
Date: Thu, 23 Aug 2018 16:57:19 +0530
Subject: [PATCH 288/334] [6917] Add ov5640 Dual Camera (CAM0 & CAM1) support

*) Enable lvs1, l17, l18 regulators for CAM0 & CAM1 ports
*) Add ov5640 dual camera device tree support
*) Make cci, camss, ov5640 as built-in modules
*) Update OV5640 register settings
*) Add CSI2 transmitter(sensor) parallel port output pixel clock (PCLK) of 96MHz

Change-Id: Iae4316400ccaef769c59e5695606b38969d763be
---
 arch/arm64/boot/dts/qcom/apq8096-openq.dtsi | 109 ++++++++++
 arch/arm64/configs/defconfig                |   7 +-
 drivers/media/i2c/ov5640.c                  | 222 ++++++++++++++++----
 kernel/configs/distro.config                |   2 +-
 4 files changed, 291 insertions(+), 49 deletions(-)

diff --git a/arch/arm64/boot/dts/qcom/apq8096-openq.dtsi b/arch/arm64/boot/dts/qcom/apq8096-openq.dtsi
index 1093a9f16841..b8806a3bcb8e 100644
--- a/arch/arm64/boot/dts/qcom/apq8096-openq.dtsi
+++ b/arch/arm64/boot/dts/qcom/apq8096-openq.dtsi
@@ -91,3 +91,112 @@
 
 	/delete-node/ user3;
 };
+&{/} {
+	rpm-glink {
+		rpm_requests {
+			pm8994-regulators {
+
+				vdd_lvs1_2-supply = <&pm8994_s4>;
+
+				pm8994_lvs1: lvs1 {
+					regulator-min-microvolt = <1800000>;
+					regulator-max-microvolt = <1800000>;
+					regulator-always-on;
+				};
+
+				l17 {
+					regulator-always-on;
+				};
+
+				l18 {
+					regulator-always-on;
+				};
+			};
+		};
+	};
+};
+
+&soc {
+	cci@a0c000 {
+		status = "ok";
+
+	camera_rear@78 {
+		compatible = "ovti,ov5640";
+		reg = <0x78>;
+
+		powerdown-gpios = <&msmgpio 26 GPIO_ACTIVE_HIGH>;
+		reset-gpios = <&msmgpio 25 GPIO_ACTIVE_LOW>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&camera_rear_default>;
+
+		clocks = <&mmcc CAMSS_MCLK0_CLK>;
+		clock-names = "xclk";
+		clock-frequency = <23880000>;
+		status = "ok";
+
+		port {
+			ov5640_ep: endpoint {
+			clock-lanes = <1>;
+			data-lanes = <0 2>;
+			remote-endpoint = <&csiphy0_ep>;
+			};
+		};
+	};
+
+
+	camera_board@3c {
+		compatible = "ovti,ov5640";
+		reg = <0x3c>;
+
+		powerdown-gpios = <&msmgpio 98 GPIO_ACTIVE_HIGH>;
+		reset-gpios = <&msmgpio 104 GPIO_ACTIVE_LOW>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&camera_board_default>;
+
+		clocks = <&mmcc CAMSS_MCLK1_CLK>;
+		clock-names = "xclk";
+		clock-frequency = <23880000>;
+		status = "ok";
+
+		port {
+			ov5640_2_ep: endpoint {
+			clock-lanes = <1>;
+			data-lanes = <0 2>;
+			remote-endpoint = <&csiphy1_ep>;
+			};
+		};
+	};
+
+	};
+
+	camss@a00000 {
+		status = "ok";
+
+		ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		port@0 {
+			reg = <0>;
+			csiphy0_ep: endpoint {
+			clock-lanes = <7>;
+			data-lanes = <0 1>;
+			remote-endpoint = <&ov5640_ep>;
+			status = "ok";
+			};
+		};
+
+		port@1 {
+			reg = <1>;
+			csiphy1_ep: endpoint {
+			clock-lanes = <7>;
+			data-lanes = <0 1>;
+			remote-endpoint = <&ov5640_2_ep>;
+			status = "ok";
+			};
+		};
+
+	};
+
+	};
+};
diff --git a/arch/arm64/configs/defconfig b/arch/arm64/configs/defconfig
index 41cd3125f62c..da5666a3827d 100644
--- a/arch/arm64/configs/defconfig
+++ b/arch/arm64/configs/defconfig
@@ -282,7 +282,7 @@ CONFIG_I2C_IMX=y
 CONFIG_I2C_MESON=y
 CONFIG_I2C_MV64XXX=y
 CONFIG_I2C_PXA=y
-CONFIG_I2C_QCOM_CCI=m
+CONFIG_I2C_QCOM_CCI=y
 CONFIG_I2C_QUP=y
 CONFIG_I2C_RK3X=y
 CONFIG_I2C_SH_MOBILE=y
@@ -370,12 +370,12 @@ CONFIG_REGULATOR_S2MPS11=y
 CONFIG_RC_CORE=m
 CONFIG_RC_DEVICES=y
 CONFIG_IR_MESON=m
-CONFIG_MEDIA_SUPPORT=m
+CONFIG_MEDIA_SUPPORT=y
 CONFIG_MEDIA_CAMERA_SUPPORT=y
 CONFIG_MEDIA_CONTROLLER=y
 CONFIG_VIDEO_V4L2_SUBDEV_API=y
 CONFIG_V4L_PLATFORM_DRIVERS=y
-CONFIG_VIDEO_QCOM_CAMSS=m
+CONFIG_VIDEO_QCOM_CAMSS=y
 CONFIG_V4L_MEM2MEM_DRIVERS=y
 CONFIG_VIDEO_SAMSUNG_S5P_JPEG=m
 CONFIG_VIDEO_SAMSUNG_S5P_MFC=m
@@ -385,6 +385,7 @@ CONFIG_VIDEO_RENESAS_VSP1=m
 CONFIG_VIDEO_QCOM_VENUS=m
 # CONFIG_MEDIA_SUBDRV_AUTOSELECT is not set
 CONFIG_VIDEO_OV5645=m
+CONFIG_VIDEO_OV5640=y
 CONFIG_VIDEO_OV7251=m
 CONFIG_DRM=m
 CONFIG_DRM_NOUVEAU=m
diff --git a/drivers/media/i2c/ov5640.c b/drivers/media/i2c/ov5640.c
index 39a2269c0bee..33872adda50b 100644
--- a/drivers/media/i2c/ov5640.c
+++ b/drivers/media/i2c/ov5640.c
@@ -163,6 +163,7 @@ struct ov5640_ctrls {
 		struct v4l2_ctrl *gain;
 	};
 	struct v4l2_ctrl *brightness;
+	struct v4l2_ctrl *pixel_clock;
 	struct v4l2_ctrl *saturation;
 	struct v4l2_ctrl *contrast;
 	struct v4l2_ctrl *hue;
@@ -564,30 +565,95 @@ static const struct reg_value ov5640_setting_15fps_PAL_720_576[] = {
 	{0x4407, 0x04, 0, 0}, {0x460b, 0x35, 0, 0}, {0x460c, 0x22, 0, 0},
 	{0x3824, 0x02, 0, 0}, {0x5001, 0xa3, 0, 0},
 };
-
 static const struct reg_value ov5640_setting_30fps_720P_1280_720[] = {
-	{0x3008, 0x42, 0, 0},
-	{0x3035, 0x21, 0, 0}, {0x3036, 0x54, 0, 0}, {0x3c07, 0x07, 0, 0},
-	{0x3c09, 0x1c, 0, 0}, {0x3c0a, 0x9c, 0, 0}, {0x3c0b, 0x40, 0, 0},
-	{0x3820, 0x41, 0, 0}, {0x3821, 0x07, 0, 0}, {0x3814, 0x31, 0, 0},
-	{0x3815, 0x31, 0, 0}, {0x3800, 0x00, 0, 0}, {0x3801, 0x00, 0, 0},
-	{0x3802, 0x00, 0, 0}, {0x3803, 0xfa, 0, 0}, {0x3804, 0x0a, 0, 0},
-	{0x3805, 0x3f, 0, 0}, {0x3806, 0x06, 0, 0}, {0x3807, 0xa9, 0, 0},
-	{0x3808, 0x05, 0, 0}, {0x3809, 0x00, 0, 0}, {0x380a, 0x02, 0, 0},
-	{0x380b, 0xd0, 0, 0}, {0x380c, 0x07, 0, 0}, {0x380d, 0x64, 0, 0},
-	{0x380e, 0x02, 0, 0}, {0x380f, 0xe4, 0, 0}, {0x3810, 0x00, 0, 0},
-	{0x3811, 0x10, 0, 0}, {0x3812, 0x00, 0, 0}, {0x3813, 0x04, 0, 0},
-	{0x3618, 0x00, 0, 0}, {0x3612, 0x29, 0, 0}, {0x3708, 0x64, 0, 0},
-	{0x3709, 0x52, 0, 0}, {0x370c, 0x03, 0, 0}, {0x3a02, 0x02, 0, 0},
-	{0x3a03, 0xe4, 0, 0}, {0x3a08, 0x01, 0, 0}, {0x3a09, 0xbc, 0, 0},
-	{0x3a0a, 0x01, 0, 0}, {0x3a0b, 0x72, 0, 0}, {0x3a0e, 0x01, 0, 0},
-	{0x3a0d, 0x02, 0, 0}, {0x3a14, 0x02, 0, 0}, {0x3a15, 0xe4, 0, 0},
-	{0x4001, 0x02, 0, 0}, {0x4004, 0x02, 0, 0}, {0x4713, 0x02, 0, 0},
-	{0x4407, 0x04, 0, 0}, {0x460b, 0x37, 0, 0}, {0x460c, 0x20, 0, 0},
-	{0x3824, 0x04, 0, 0}, {0x5001, 0x83, 0, 0}, {0x4005, 0x1a, 0, 0},
-	{0x3008, 0x02, 0, 0}, {0x3503, 0,    0, 0},
+	/*@@ MIPI_2lane_720P(YUV) 60fps*/
+	{0x3103, 0x11, 0, 0},	{0x3008, 0x82, 0, 0},	{0x3008, 0x42, 0, 0},
+	{0x3103, 0x03, 0, 0},	{0x3017, 0x00, 0, 0},	{0x3018, 0x00, 0, 0},
+	{0x3034, 0x18, 0, 0},	{0x3035, 0x11, 0, 0},	{0x3036, 0x54, 0, 0},
+	{0x3037, 0x13, 0, 0},	{0x3108, 0x01, 0, 0},	{0x3630, 0x36, 0, 0},
+	{0x3631, 0x0e, 0, 0},	{0x3632, 0xe2, 0, 0},	{0x3633, 0x12, 0, 0},
+	{0x3621, 0xe0, 0, 0},	{0x3704, 0xa0, 0, 0},	{0x3703, 0x5a, 0, 0},
+	{0x3715, 0x78, 0, 0},	{0x3717, 0x01, 0, 0},	{0x370b, 0x60, 0, 0},
+	{0x3705, 0x1a, 0, 0},	{0x3905, 0x02, 0, 0},	{0x3906, 0x10, 0, 0},
+	{0x3901, 0x0a, 0, 0},	{0x3731, 0x12, 0, 0},	{0x3600, 0x08, 0, 0},
+	{0x3601, 0x33, 0, 0},	{0x302d, 0x60, 0, 0},	{0x3620, 0x52, 0, 0},
+	{0x371b, 0x20, 0, 0},	{0x471c, 0x50, 0, 0},	{0x3a13, 0x43, 0, 0},
+	{0x3a18, 0x00, 0, 0},	{0x3a19, 0xf8, 0, 0},	{0x3635, 0x13, 0, 0},
+	{0x3636, 0x03, 0, 0},	{0x3634, 0x40, 0, 0},	{0x3622, 0x01, 0, 0},
+	{0x3c01, 0x34, 0, 0},	{0x3c04, 0x28, 0, 0},	{0x3c05, 0x98, 0, 0},
+	{0x3c06, 0x00, 0, 0},	{0x3c07, 0x07, 0, 0},	{0x3c08, 0x00, 0, 0},
+	{0x3c09, 0x1c, 0, 0},	{0x3c0a, 0x9c, 0, 0},	{0x3c0b, 0x40, 0, 0},
+	{0x3820, 0x41, 0, 0},	{0x3821, 0x07, 0, 0},	{0x3814, 0x31, 0, 0},
+	{0x3815, 0x31, 0, 0},	{0x3800, 0x00, 0, 0},	{0x3801, 0x00, 0, 0},
+	{0x3802, 0x00, 0, 0},	{0x3803, 0xfa, 0, 0},	{0x3804, 0x0a, 0, 0},
+	{0x3805, 0x3f, 0, 0},	{0x3806, 0x06, 0, 0},	{0x3807, 0xa9, 0, 0},
+	{0x3808, 0x05, 0, 0},	{0x3809, 0x00, 0, 0},	{0x380a, 0x02, 0, 0},
+	{0x380b, 0xd0, 0, 0},	{0x380c, 0x07, 0, 0},	{0x380d, 0x64, 0, 0},
+	{0x380e, 0x02, 0, 0},	{0x380f, 0xe4, 0, 0},	{0x3810, 0x00, 0, 0},
+	{0x3811, 0x10, 0, 0},	{0x3812, 0x00, 0, 0},	{0x3813, 0x04, 0, 0},
+	{0x3618, 0x00, 0, 0},	{0x3612, 0x29, 0, 0},	{0x3708, 0x64, 0, 0},
+	{0x3709, 0x52, 0, 0},	{0x370c, 0x03, 0, 0},	{0x3a02, 0x02, 0, 0},
+	{0x3a03, 0xe4, 0, 0},	{0x3a08, 0x01, 0, 0},	{0x3a09, 0xbc, 0, 0},
+	{0x3a0a, 0x01, 0, 0},	{0x3a0b, 0x72, 0, 0},	{0x3a0e, 0x01, 0, 0},
+	{0x3a0d, 0x02, 0, 0},	{0x3a14, 0x02, 0, 0},	{0x3a15, 0xe4, 0, 0},
+	{0x4001, 0x02, 0, 0},	{0x4004, 0x02, 0, 0},	{0x3000, 0x00, 0, 0},
+	{0x3002, 0x1c, 0, 0},	{0x3004, 0xff, 0, 0},	{0x3006, 0xc3, 0, 0},
+	{0x300e, 0x45, 0, 0},	{0x302e, 0x08, 0, 0},	{0x4300, 0x32, 0, 0},
+	{0x501f, 0x00, 0, 0},	{0x4713, 0x02, 0, 0},	{0x4407, 0x04, 0, 0},
+	{0x440e, 0x00, 0, 0},	{0x460b, 0x37, 0, 0},	{0x460c, 0x20, 0, 0},
+	{0x4800, 0x24, 0, 0},	{0x4827, 0x16, 0, 0},	{0x4837, 0x0a, 0, 0},
+	{0x3824, 0x04, 0, 0},	{0x5000, 0xa7, 0, 0},	{0x5001, 0x83, 0, 0},
+	{0x5180, 0xff, 0, 0},	{0x5181, 0xf2, 0, 0},	{0x5182, 0x00, 0, 0},
+	{0x5183, 0x14, 0, 0},	{0x5185, 0x24, 0, 0},	{0x5186, 0x09, 0, 0},
+	{0x5187, 0x09, 0, 0},	{0x5188, 0x09, 0, 0},	{0x5189, 0x75, 0, 0},
+	{0x518a, 0x54, 0, 0},	{0x518b, 0xe0, 0, 0},	{0x518c, 0xb2, 0, 0},
+	{0x518d, 0x42, 0, 0},	{0x518e, 0x3d, 0, 0},	{0x518f, 0x56, 0, 0},
+	{0x5190, 0x46, 0, 0},	{0x5191, 0xf8, 0, 0},	{0x5192, 0x04, 0, 0},
+	{0x5193, 0x70, 0, 0},	{0x5194, 0xf0, 0, 0},	{0x5195, 0xf0, 0, 0},
+	{0x5196, 0x03, 0, 0},	{0x5197, 0x01, 0, 0},	{0x5198, 0x04, 0, 0},
+	{0x5199, 0x12, 0, 0},	{0x519a, 0x04, 0, 0},	{0x519b, 0x00, 0, 0},
+	{0x519c, 0x06, 0, 0},	{0x519d, 0x82, 0, 0},	{0x519e, 0x38, 0, 0},
+	{0x5381, 0x1e, 0, 0},	{0x5382, 0x5b, 0, 0},	{0x5383, 0x08, 0, 0},
+	{0x5384, 0x0a, 0, 0},	{0x5385, 0x7e, 0, 0},	{0x5386, 0x88, 0, 0},
+	{0x5387, 0x7c, 0, 0},	{0x5388, 0x6c, 0, 0},	{0x5389, 0x10, 0, 0},
+	{0x538a, 0x01, 0, 0},	{0x538b, 0x98, 0, 0},	{0x5300, 0x08, 0, 0},
+	{0x5301, 0x30, 0, 0},	{0x5302, 0x10, 0, 0},	{0x5303, 0x00, 0, 0},
+	{0x5304, 0x08, 0, 0},	{0x5305, 0x30, 0, 0},	{0x5306, 0x08, 0, 0},
+	{0x5307, 0x16, 0, 0},	{0x5309, 0x08, 0, 0},	{0x530a, 0x30, 0, 0},
+	{0x530b, 0x04, 0, 0},	{0x530c, 0x06, 0, 0},	{0x5480, 0x01, 0, 0},
+	{0x5481, 0x08, 0, 0},	{0x5482, 0x14, 0, 0},	{0x5483, 0x28, 0, 0},
+	{0x5484, 0x51, 0, 0},	{0x5485, 0x65, 0, 0},	{0x5486, 0x71, 0, 0},
+	{0x5487, 0x7d, 0, 0},	{0x5488, 0x87, 0, 0},	{0x5489, 0x91, 0, 0},
+	{0x548a, 0x9a, 0, 0},	{0x548b, 0xaa, 0, 0},	{0x548c, 0xb8, 0, 0},
+	{0x548d, 0xcd, 0, 0},	{0x548e, 0xdd, 0, 0},	{0x548f, 0xea, 0, 0},
+	{0x5490, 0x1d, 0, 0},	{0x5580, 0x02, 0, 0},	{0x5583, 0x40, 0, 0},
+	{0x5584, 0x10, 0, 0},	{0x5589, 0x10, 0, 0},	{0x558a, 0x00, 0, 0},
+	{0x558b, 0xf8, 0, 0},	{0x5800, 0x23, 0, 0},	{0x5801, 0x14, 0, 0},
+	{0x5802, 0x0f, 0, 0},	{0x5803, 0x0f, 0, 0},	{0x5804, 0x12, 0, 0},
+	{0x5805, 0x26, 0, 0},	{0x5806, 0x0c, 0, 0},	{0x5807, 0x08, 0, 0},
+	{0x5808, 0x05, 0, 0},	{0x5809, 0x05, 0, 0},	{0x580a, 0x08, 0, 0},
+	{0x580b, 0x0d, 0, 0},	{0x580c, 0x08, 0, 0},	{0x580d, 0x03, 0, 0},
+	{0x580e, 0x00, 0, 0},	{0x580f, 0x00, 0, 0},	{0x5810, 0x03, 0, 0},
+	{0x5811, 0x09, 0, 0},	{0x5812, 0x07, 0, 0},	{0x5813, 0x03, 0, 0},
+	{0x5814, 0x00, 0, 0},	{0x5815, 0x01, 0, 0},	{0x5816, 0x03, 0, 0},
+	{0x5817, 0x08, 0, 0},	{0x5818, 0x0d, 0, 0},	{0x5819, 0x08, 0, 0},
+	{0x581a, 0x05, 0, 0},	{0x581b, 0x06, 0, 0},	{0x581c, 0x08, 0, 0},
+	{0x581d, 0x0e, 0, 0},	{0x581e, 0x29, 0, 0},	{0x581f, 0x17, 0, 0},
+	{0x5820, 0x11, 0, 0},	{0x5821, 0x11, 0, 0},	{0x5822, 0x15, 0, 0},
+	{0x5823, 0x28, 0, 0},	{0x5824, 0x46, 0, 0},	{0x5825, 0x26, 0, 0},
+	{0x5826, 0x08, 0, 0},	{0x5827, 0x26, 0, 0},	{0x5828, 0x64, 0, 0},
+	{0x5829, 0x26, 0, 0},	{0x582a, 0x24, 0, 0},	{0x582b, 0x22, 0, 0},
+	{0x582c, 0x24, 0, 0},	{0x582d, 0x24, 0, 0},	{0x582e, 0x06, 0, 0},
+	{0x582f, 0x22, 0, 0},	{0x5830, 0x40, 0, 0},	{0x5831, 0x42, 0, 0},
+	{0x5832, 0x24, 0, 0},	{0x5833, 0x26, 0, 0},	{0x5834, 0x24, 0, 0},
+	{0x5835, 0x22, 0, 0},	{0x5836, 0x22, 0, 0},	{0x5837, 0x26, 0, 0},
+	{0x5838, 0x44, 0, 0},	{0x5839, 0x24, 0, 0},	{0x583a, 0x26, 0, 0},
+	{0x583b, 0x28, 0, 0},	{0x583c, 0x42, 0, 0},	{0x583d, 0xce, 0, 0},
+	{0x5025, 0x00, 0, 0},	{0x3a0f, 0x30, 0, 0},	{0x3a10, 0x28, 0, 0},
+	{0x3a1b, 0x30, 0, 0},	{0x3a1e, 0x26, 0, 0},	{0x3a11, 0x60, 0, 0},
+	{0x5308, 0x10, 0, 0},	{0x5306, 0x1f, 0, 0},	{0x3a1f, 0x14, 0, 0},
+	{0x3008, 0x02, 0, 0},
 };
-
 static const struct reg_value ov5640_setting_15fps_720P_1280_720[] = {
 	{0x3035, 0x41, 0, 0}, {0x3036, 0x54, 0, 0}, {0x3c07, 0x07, 0, 0},
 	{0x3c09, 0x1c, 0, 0}, {0x3c0a, 0x9c, 0, 0}, {0x3c0b, 0x40, 0, 0},
@@ -681,29 +747,91 @@ static const struct reg_value ov5640_setting_15fps_1080P_1920_1080[] = {
 	{0x460b, 0x37, 0, 0}, {0x460c, 0x20, 0, 0}, {0x3824, 0x04, 0, 0},
 	{0x4005, 0x1a, 0, 0}, {0x3008, 0x02, 0, 0}, {0x3503, 0, 0, 0},
 };
-
 static const struct reg_value ov5640_setting_15fps_QSXGA_2592_1944[] = {
-	{0x3820, 0x40, 0, 0}, {0x3821, 0x06, 0, 0},
-	{0x3035, 0x21, 0, 0}, {0x3036, 0x54, 0, 0}, {0x3c07, 0x08, 0, 0},
-	{0x3c09, 0x1c, 0, 0}, {0x3c0a, 0x9c, 0, 0}, {0x3c0b, 0x40, 0, 0},
-	{0x3820, 0x40, 0, 0}, {0x3821, 0x06, 0, 0}, {0x3814, 0x11, 0, 0},
-	{0x3815, 0x11, 0, 0}, {0x3800, 0x00, 0, 0}, {0x3801, 0x00, 0, 0},
-	{0x3802, 0x00, 0, 0}, {0x3803, 0x00, 0, 0}, {0x3804, 0x0a, 0, 0},
-	{0x3805, 0x3f, 0, 0}, {0x3806, 0x07, 0, 0}, {0x3807, 0x9f, 0, 0},
-	{0x3808, 0x0a, 0, 0}, {0x3809, 0x20, 0, 0}, {0x380a, 0x07, 0, 0},
-	{0x380b, 0x98, 0, 0}, {0x380c, 0x0b, 0, 0}, {0x380d, 0x1c, 0, 0},
-	{0x380e, 0x07, 0, 0}, {0x380f, 0xb0, 0, 0}, {0x3810, 0x00, 0, 0},
-	{0x3811, 0x10, 0, 0}, {0x3812, 0x00, 0, 0}, {0x3813, 0x04, 0, 0},
-	{0x3618, 0x04, 0, 0}, {0x3612, 0x29, 0, 0}, {0x3708, 0x21, 0, 0},
-	{0x3709, 0x12, 0, 0}, {0x370c, 0x00, 0, 0}, {0x3a02, 0x03, 0, 0},
-	{0x3a03, 0xd8, 0, 0}, {0x3a08, 0x01, 0, 0}, {0x3a09, 0x27, 0, 0},
-	{0x3a0a, 0x00, 0, 0}, {0x3a0b, 0xf6, 0, 0}, {0x3a0e, 0x03, 0, 0},
-	{0x3a0d, 0x04, 0, 0}, {0x3a14, 0x03, 0, 0}, {0x3a15, 0xd8, 0, 0},
-	{0x4001, 0x02, 0, 0}, {0x4004, 0x06, 0, 0}, {0x4713, 0x03, 0, 0},
-	{0x4407, 0x04, 0, 0}, {0x460b, 0x35, 0, 0}, {0x460c, 0x22, 0, 0},
-	{0x3824, 0x02, 0, 0}, {0x5001, 0x83, 0, 70},
+	{0x3103, 0x11, 0, 0},	{0x3008, 0x82},	{0x3008, 0x42, 0, 0},
+	{0x3103, 0x03, 0, 0},	{0x3017, 0x00},	{0x3018, 0x00, 0, 0},
+	{0x3034, 0x18, 0, 0},	{0x3035, 0x11},	{0x3036, 0x54, 0, 0},
+	{0x3037, 0x13, 0, 0},	{0x3108, 0x01},	{0x3630, 0x36, 0, 0},
+	{0x3631, 0x0e, 0, 0},	{0x3632, 0xe2},	{0x3633, 0x12, 0, 0},
+	{0x3621, 0xe0, 0, 0},	{0x3704, 0xa0},	{0x3703, 0x5a, 0, 0},
+	{0x3715, 0x78, 0, 0},	{0x3717, 0x01},	{0x370b, 0x60, 0, 0},
+	{0x3705, 0x1a, 0, 0},	{0x3905, 0x02},	{0x3906, 0x10, 0, 0},
+	{0x3901, 0x0a, 0, 0},	{0x3731, 0x12},	{0x3600, 0x08, 0, 0},
+	{0x3601, 0x33, 0, 0},	{0x302d, 0x60},	{0x3620, 0x52, 0, 0},
+	{0x371b, 0x20, 0, 0},	{0x471c, 0x50},	{0x3a13, 0x43, 0, 0},
+	{0x3a18, 0x00, 0, 0},	{0x3a19, 0xf8},	{0x3635, 0x13, 0, 0},
+	{0x3636, 0x03, 0, 0},	{0x3634, 0x40},	{0x3622, 0x01, 0, 0},
+	{0x3c01, 0x34, 0, 0},	{0x3c04, 0x28},	{0x3c05, 0x98, 0, 0},
+	{0x3c06, 0x00, 0, 0},	{0x3c07, 0x07},	{0x3c08, 0x00, 0, 0},
+	{0x3c09, 0x1c, 0, 0},	{0x3c0a, 0x9c},	{0x3c0b, 0x40, 0, 0},
+	{0x3820, 0x40, 0, 0},	{0x3821, 0x06},	{0x3814, 0x11, 0, 0},
+	{0x3815, 0x11, 0, 0},	{0x3800, 0x00},	{0x3801, 0x00, 0, 0},
+	{0x3802, 0x00, 0, 0},	{0x3803, 0x00},	{0x3804, 0x0a, 0, 0},
+	{0x3805, 0x3f, 0, 0},	{0x3806, 0x07},	{0x3807, 0x9f, 0, 0},
+	{0x3808, 0x0a, 0, 0},	{0x3809, 0x20},	{0x380a, 0x07, 0, 0},
+	{0x380b, 0x98, 0, 0},	{0x380c, 0x0b},	{0x380d, 0x1c, 0, 0},
+	{0x380e, 0x07, 0, 0}, 	{0x380f, 0xb0},	{0x3810, 0x00, 0, 0},
+	{0x3811, 0x10, 0, 0},	{0x3812, 0x00},	{0x3813, 0x04, 0, 0},
+	{0x3618, 0x04, 0, 0},	{0x3612, 0x2b},	{0x3708, 0x64, 0, 0},
+	{0x3709, 0x12, 0, 0},	{0x370c, 0x00},	{0x3a02, 0x07, 0, 0},
+	{0x3a03, 0xb0, 0, 0},	{0x3a08, 0x01},	{0x3a09, 0x27, 0, 0},
+	{0x3a0a, 0x00, 0, 0},	{0x3a0b, 0xf6},	{0x3a0e, 0x06, 0, 0},
+	{0x3a0d, 0x08, 0, 0},	{0x3a14, 0x07},	{0x3a15, 0xb0, 0, 0},
+	{0x4001, 0x02, 0, 0},	{0x4004, 0x06},	{0x3000, 0x00, 0, 0},
+	{0x3002, 0x1c, 0, 0},	{0x3004, 0xff},	{0x3006, 0xc3, 0, 0},
+	{0x300e, 0x45, 0, 0},	{0x302e, 0x08},	{0x4300, 0x32, 0, 0},
+	{0x4837, 0x0a, 0, 0},	{0x501f, 0x00},	{0x440e, 0x00, 0, 0},
+	{0x4800, 0x24, 0, 0},	{0x5000, 0xa7},	{0x5001, 0x83, 0, 0},
+	{0x5180, 0xff, 0, 0},	{0x5181, 0xf2},	{0x5182, 0x00, 0, 0},
+	{0x5183, 0x14, 0, 0},	{0x5184, 0x25},	{0x5185, 0x24, 0, 0},
+	{0x5186, 0x09, 0, 0},	{0x5187, 0x09},	{0x5188, 0x09, 0, 0},
+	{0x5189, 0x75, 0, 0},	{0x518a, 0x54},	{0x518b, 0xe0, 0, 0},
+	{0x518c, 0xb2, 0, 0},	{0x518d, 0x42},	{0x518e, 0x3d, 0, 0},
+	{0x518f, 0x56, 0, 0},	{0x5190, 0x46},	{0x5191, 0xf8, 0, 0},
+	{0x5192, 0x04, 0, 0},	{0x5193, 0x70},	{0x5194, 0xf0, 0, 0},
+	{0x5195, 0xf0, 0, 0},	{0x5196, 0x03},	{0x5197, 0x01, 0, 0},
+	{0x5198, 0x04, 0, 0},	{0x5199, 0x12},	{0x519a, 0x04, 0, 0},
+	{0x519b, 0x00, 0, 0},	{0x519c, 0x06},	{0x519d, 0x82, 0, 0},
+	{0x519e, 0x38, 0, 0},	{0x5381, 0x1e},	{0x5382, 0x5b, 0, 0},
+	{0x5383, 0x08, 0, 0},	{0x5384, 0x0a},	{0x5385, 0x7e, 0, 0},
+	{0x5386, 0x88, 0, 0},	{0x5387, 0x7c},	{0x5388, 0x6c, 0, 0},
+	{0x5389, 0x10, 0, 0},	{0x538a, 0x01},	{0x538b, 0x98, 0, 0},
+	{0x5300, 0x08, 0, 0},	{0x5301, 0x30},	{0x5302, 0x10, 0, 0},
+	{0x5303, 0x00, 0, 0},	{0x5304, 0x08},	{0x5305, 0x30, 0, 0},
+	{0x5306, 0x08, 0, 0},	{0x5307, 0x16},	{0x5309, 0x08, 0, 0},
+	{0x530a, 0x30, 0, 0},	{0x530b, 0x04},	{0x530c, 0x06, 0, 0},
+	{0x5480, 0x01, 0, 0},	{0x5481, 0x08},	{0x5482, 0x14, 0, 0},
+	{0x5483, 0x28, 0, 0},	{0x5484, 0x51},	{0x5485, 0x65, 0, 0},
+	{0x5486, 0x71, 0, 0},	{0x5487, 0x7d},	{0x5488, 0x87, 0, 0},
+	{0x5489, 0x91, 0, 0},	{0x548a, 0x9a},	{0x548b, 0xaa, 0, 0},
+	{0x548c, 0xb8, 0, 0},	{0x548d, 0xcd},	{0x548e, 0xdd, 0, 0},
+	{0x548f, 0xea, 0, 0},	{0x5490, 0x1d},	{0x5580, 0x02, 0, 0},
+	{0x5583, 0x40, 0, 0},	{0x5584, 0x10},	{0x5589, 0x10, 0, 0},
+	{0x558a, 0x00, 0, 0},	{0x558b, 0xf8},	{0x5800, 0x23, 0, 0},
+	{0x5801, 0x14, 0, 0},	{0x5802, 0x0f},	{0x5803, 0x0f, 0, 0},
+	{0x5804, 0x12, 0, 0},	{0x5805, 0x26},	{0x5806, 0x0c, 0, 0},
+	{0x5807, 0x08, 0, 0},	{0x5808, 0x05},	{0x5809, 0x05, 0, 0},
+	{0x580a, 0x08, 0, 0},	{0x580b, 0x0d},	{0x580c, 0x08, 0, 0},
+	{0x580d, 0x03, 0, 0},	{0x580e, 0x00},	{0x580f, 0x00, 0, 0},
+	{0x5810, 0x03, 0, 0},	{0x5811, 0x09},	{0x5812, 0x07, 0, 0},
+	{0x5813, 0x03, 0, 0},	{0x5814, 0x00},	{0x5815, 0x01, 0, 0},
+	{0x5816, 0x03, 0, 0},	{0x5817, 0x08},	{0x5818, 0x0d, 0, 0},
+	{0x5819, 0x08, 0, 0},	{0x581a, 0x05},	{0x581b, 0x06, 0, 0},
+	{0x581c, 0x08, 0, 0},	{0x581d, 0x0e},	{0x581e, 0x29, 0, 0},
+	{0x581f, 0x17, 0, 0},	{0x5820, 0x11},	{0x5821, 0x11, 0, 0},
+	{0x5822, 0x15, 0, 0},	{0x5823, 0x28},	{0x5824, 0x46, 0, 0},
+	{0x5825, 0x26, 0, 0},	{0x5826, 0x08},	{0x5827, 0x26, 0, 0},
+	{0x5828, 0x64, 0, 0},	{0x5829, 0x26},	{0x582a, 0x24, 0, 0},
+	{0x582b, 0x22, 0, 0},	{0x582c, 0x24},	{0x582d, 0x24, 0, 0},
+	{0x582e, 0x06, 0, 0},	{0x582f, 0x22},	{0x5830, 0x40, 0, 0},
+	{0x5831, 0x42, 0, 0},	{0x5832, 0x24},	{0x5833, 0x26, 0, 0},
+	{0x5834, 0x24, 0, 0},	{0x5835, 0x22},	{0x5836, 0x22, 0, 0},
+	{0x5837, 0x26, 0, 0},	{0x5838, 0x44},	{0x5839, 0x24, 0, 0},
+	{0x583a, 0x26, 0, 0},	{0x583b, 0x28},	{0x583c, 0x42, 0, 0},
+	{0x583d, 0xce, 0, 0},	{0x5025, 0x00},	{0x3a0f, 0x30, 0, 0},
+	{0x3a10, 0x28, 0, 0},	{0x3a1b, 0x30},	{0x3a1e, 0x26, 0, 0},
+	{0x3a11, 0x60, 0, 0},	{0x3a1f, 0x14},	{0x3008, 0x02, 0, 0},
 };
-
 /* power-on sensor init reg table */
 static const struct ov5640_mode_info ov5640_mode_init_data = {
 	0, SUBSAMPLING, 640, 480, ov5640_init_setting_30fps_VGA,
@@ -765,7 +893,10 @@ ov5640_mode_data[OV5640_NUM_FRAMERATES][OV5640_NUM_MODES] = {
 		{OV5640_MODE_1080P_1920_1080, SCALING, 1920, 1080,
 		 ov5640_setting_30fps_1080P_1920_1080,
 		 ARRAY_SIZE(ov5640_setting_30fps_1080P_1920_1080)},
-		{OV5640_MODE_QSXGA_2592_1944, -1, 0, 0, NULL, 0},
+		//{OV5640_MODE_QSXGA_2592_1944, -1, 0, 0, NULL, 0},
+		{OV5640_MODE_QSXGA_2592_1944, SCALING, 2592, 1944,
+		 ov5640_setting_15fps_QSXGA_2592_1944,
+		ARRAY_SIZE(ov5640_setting_15fps_QSXGA_2592_1944)},
 	},
 };
 
@@ -2008,7 +2139,8 @@ static int ov5640_init_controls(struct ov5640_dev *sensor)
 					     0, 1, 1, 1);
 	ctrls->gain = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_GAIN,
 					0, 1023, 1, 0);
-
+	ctrls->pixel_clock = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_PIXEL_RATE,
+					1, INT_MAX, 1, 96000000);
 	ctrls->saturation = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_SATURATION,
 					      0, 255, 1, 64);
 	ctrls->hue = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_HUE,
diff --git a/kernel/configs/distro.config b/kernel/configs/distro.config
index b503edb5b8e8..20a167488891 100644
--- a/kernel/configs/distro.config
+++ b/kernel/configs/distro.config
@@ -1,5 +1,5 @@
 # USB camera
-CONFIG_MEDIA_SUPPORT=m
+CONFIG_MEDIA_SUPPORT=y
 CONFIG_MEDIA_CAMERA_SUPPORT=y
 CONFIG_MEDIA_USB_SUPPORT=y
 CONFIG_USB_VIDEO_CLASS=m
-- 
2.24.1

